### ISSUE TITLE
Incompatibility between kvantum and spyder

### ISSUE THREAD

**Issue by Ticonderoga** [spyder-ide/spyder#16826]
It's a followup of issue #11623. It's impossible to launch spyder with kvantum on and I tried with the github version and the arch version and the conda one. An easy  fix is to add these two lines in /usr/bin/spyder : 

import os
os.environ['QT_STYLE_OVERRIDE'] = 'adwaita'

since april I manually add these two lines and I can't understand the bug. It's also mentionned here 

https://githubmemory.com/repo/tsujan/Kvantum/issues/563

Hope that my fix will help !

_Originally posted by @Ticonderoga in https://github.com/spyder-ide/spyder/issues/11623#issuecomment-968715609_

**Comment by ccordoba12** [comment]
Hey @Ticonderoga, thanks for reporting. I just don't know what we could do in this case because it seems Kvantum doesn't allow Spyder to set its own Qt style but instead it wants to impose its own, which we don't want to allow because Spyder doesn't work anymore (i.e. look well at all) with the system style.

So the question is: why an application can't set its own style under Kvantum?

**Comment by Ticonderoga** [comment]
Unfortunately I have no clue about using a custom style with Kvantum. For now I use the git version and I apply my patch each time.

**Comment by ccordoba12** [comment]
Ok, could you post a screenshot to see how Spyder looks with your patch? Thanks!

**Comment by Ticonderoga** [comment]
Actually I think it looks the same with or without the patch. The only difference is that spyder is not "stuck" after the startup. 

![2021-11-21_00-11](https://user-images.githubusercontent.com/4589148/142743492-d7d06690-016c-409d-ba53-d8aa4e949549.png)

For the record here is my spyder.ini

[spyder_ini.txt](https://github.com/spyder-ide/spyder/files/7575483/spyder_ini.txt)

My main machine is on Arch and I use the package spyder-git from aur (https://aur.archlinux.org/packages/spyder-git/) I tried - with success - my patch on 3 other PC with Manjaro and the standard spyder package (https://archlinux.org/packages/community/any/spyder/)

Hope it will help. I could try to start with the debug mode if you want ...

**Comment by ccordoba12** [comment]
> Actually I think it looks the same with or without the patch

Yes, it does. I thought it'd inherit the Kvantum theme with the patch (that's why I asked you for the screenshot).

> Hope it will help. I could try to start with the debug mode if you want ...

Thanks! I think you've done enough and the solution should be relatively simple. Could you post your patch here (for the record) and also ping the Kvantum maintainer?

My idea is to detect when Spyder is running under Kvantum (by detecting an environment variable or something similar) and then apply your patch. The thing is I wouldn't like to apply it unconditionally because it could introduce other problems.

**Comment by qouoq** [comment]
I tried to make a minimum reproduction case, and got the following result:

<details>
<summary>console_freeze.py</summary>

```python {.line-numbers}
#!/usr/bin/env python3

from PyQt5 import QtWidgets, QtCore
# from PySide2 import QtWidgets, QtCore
# from qtconsole.rich_jupyter_widget import RichJupyterWidget as TestWidget

class TestWidget(QtWidgets.QWidget):

    def __init__(self):
        super().__init__()

        layout = QtWidgets.QStackedLayout(self)
        self._control = self._create_control()
        layout.addWidget(self._control)

        self._widget = QtWidgets.QWidget()
        self._filter_resize = False

    def eventFilter(self, obj, event):
        """ Reimplemented to ensure a console-like behavior in the underlying
            text widgets.
        """
        etype = event.type()
        # Manually adjust the scrollbars *after* a resize event is dispatched.
        if etype == QtCore.QEvent.Resize:
            print(self._filter_resize)
            if not self._filter_resize:
                self._filter_resize = True
                QtWidgets.QApplication.instance().sendEvent(obj, event)
                self._adjust_scrollbars()
                self._filter_resize = False
                return True

        return super().eventFilter(obj, event)

    def _create_control(self):
        """ Creates and connects the underlying text widget.
        """
        control = QtWidgets.QTextEdit()

        # Install event filters. The filter on the viewport is needed for
        # mouse events.
        control.installEventFilter(self)
        control.viewport().installEventFilter(self)

        # Hijack the document size change signal to prevent Qt from adjusting
        # the viewport's scrollbar. We are relying on an implementation detail
        # of Q(Plain)TextEdit here, which is potentially dangerous, but without
        # this functionality we cannot create a nice terminal interface.
        layout = control.document().documentLayout()
        layout.documentSizeChanged.disconnect()
        layout.documentSizeChanged.connect(self._adjust_scrollbars)

        control.setVerticalScrollBarPolicy(QtCore.Qt.ScrollBarAlwaysOn)
        return control

    def _adjust_scrollbars(self):
        """ Expands the vertical scrollbar beyond the range set by Qt.
        """
        # This code is adapted from _q_adjustScrollbars in qplaintextedit.cpp
        # and qtextedit.cpp.
        document = self._control.document()
        scrollbar = self._control.verticalScrollBar()
        viewport_height = self._control.viewport().height()

        # QTextEdit does not do line-based layout and blocks will not in
        # general have the same height. Therefore it does not make sense to
        # attempt to scroll in line height increments.
        maximum = document.size().height()
        step = viewport_height

        diff = maximum - scrollbar.maximum()
        scrollbar.setRange(0, round(maximum))
        scrollbar.setPageStep(round(step))

if __name__ == "__main__":
    app = QtWidgets.QApplication([])
    app.setStyleSheet('QTextEdit { border: 1px; }')
    widget = TestWidget()
    widget.show()
    app.exec_()
```
</details>

This is basically a `qtconsole.console_widget` stripped down as much as I could, with no other dependency than Qt and Python Qt bindings.

At the top, two commented imports:
- `import RichJupyterWidget` to reproduce with a standard qtconsole. To try it, the full class TestWidget should be commented while uncommenting the import.
- `import PySide2` to check that the behavior is the same with PySide2 and PyQt5, which it is.

### The grueling details

The code used in `eventFilter`, `_create_control` and `_adjust_scollbars` leads in some situations to an infinite loop of `_filter_resize` oscillating between False and True that makes the application busy enough that it appears frozen.
_(The `print _filter_resize` in `eventFilter` helps making it easy to observe.)_

The necessary and sufficient condition when using Kvantum as the Qt theme engine is the `app.setStyleSheet` line at the bottom in `__main__`. The particular css used is likely not important; I first thought that it was due to some css from QDarkStyleSheet and tried to whittle at it, ended up with a dozen lines that each would trigger the bug.

The line at the end of `_create_control`, that `setVerticalScrollBarPolicy(ScrollBarAlwaysOn)`, is of no consequence with Kvantum. Commenting it out however triggers the same oscillation of `_filter_resize` and unresponsiveness when not using Kvantum.

One last remark is that independently of the `VerticalScrollBarPolicy`, Kvantum, freezing or not, shows a vertical bar, while non-Kvantum Qt doesn't show a bar without ScrollBarAlwaysOn (and that's when it freezes). So my guess would be that the crux of the matter is somehow about the directives given to that scrollbar.

### Other comments

If you go with the solution you discussed above, don't use adwaita as the fallback theme, on my computer it wouldn't work for example, adwaita is not installed. You could use Fusion for example, which I think comes out of the box with Qt.

If you think it may be useful to report these findings to the Kvantum maintainer, let me know, I'd be glad to do.

There may be a Qt bug hiding in there, but considering the comment from the qtconsole code that 'this rely on an implementation detail', I'm not so optimistic about the idea. Besides, I imagine it's possible to get into some Qt debugging for further insight, but I have no idea how to go this way..

**Comment by ccordoba12** [comment]
@qouoq, thanks a lot for investigating the problem in such depth! You said

> If you go with the solution you discussed above, don't use adwaita as the fallback theme, on my computer it wouldn't work for example, adwaita is not installed. You could use Fusion for example, which I think comes out of the box with Qt.

Sure, no problem with that.

> If you think it may be useful to report these findings to the Kvantum maintainer, let me know, I'd be glad to do.

Yeah, I think that'd be great. The only thing we need from them is an environment variable set, so we can read it on our side and apply the patch mentioned by @Ticonderoga only when Kvantum is running.

**Comment by Ticonderoga** [comment]
With version  6.1.0.3 the patch is still working flawlessly ...

**Comment by ccordoba12** [comment]
@Ticonderoga, you mean the two lines you posted in your initial message?

> import os
os.environ['QT_STYLE_OVERRIDE'] = 'adwaita'

We could easily add them, but how can we detect that `kvantum` is being used to do that?
